mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp

```diff
@@ -334,43 +330,84 @@ static TransformingForOp createGlobalLoadLoop(PatternRewriter &b, Location loc,
       ArrayRef<Attribute>{matrixToTensor, b.getArrayAttr({})},
       /*bounds=*/ArrayRef<int64_t>{1, 1, 1, dataPerThread},
       /*strides=*/ArrayRef<int64_t>{1, 1, 1, vectorLen}, forceUnroll,
-      /*useIndexDiffs=*/true, resultInit);
+      /*useIndexDiffs=*/true);
   {
     PatternRewriter::InsertionGuard outerGuard(b);
-    b.setInsertionPointToEnd(outerLoop.getBody());
+    b.setInsertionPointToStart(outerLoop.getBody());
+
     Value loaded = b.create<GlobalLoadOp>(
         loc, loadType, tensor, outerLoop.getValidity(/*domain=*/0),
         outerLoop.getLowerCoords(/*domain=*/0));
-    auto innerLoop = b.create<TransformingForOp>(
-        loc,
-        ArrayRef<ValueRange>{zero,
-                             outerLoop.getLowerCoords(/*domain=*/1).back()},
-        ArrayRef<Attribute>{b.getArrayAttr({}), vectorMap},
-        /*bounds=*/ArrayRef<int64_t>{vectorLen},
-        /*strides=*/ArrayRef<int64_t>{1}, forceUnroll, /*useIndexDiffs=*/true,
-        outerLoop.getIterArgs()[0]);
-    {
-      PatternRewriter::InsertionGuard innerGuard(b);
-      b.setInsertionPointToEnd(innerLoop.getBody());
-      Value loadElem =
-          vectorLen == 1
-              ? loaded
-              : b.create<vector::ExtractElementOp>(
-                    loc, loaded, innerLoop.getLowerCoords(/*domain=*/0)[0]);
-      Value toYieldInner = dataPerThread == 1
-                               ? loadElem
-                               : b.create<vector::InsertElementOp>(
-                                     loc, loadElem, innerLoop.getIterArgs()[0],
-                                     innerLoop.getLowerCoords(/*domain=*/1)[0]);
-      b.create<rock::YieldOp>(loc, toYieldInner);
-    }
-    b.create<rock::YieldOp>(loc, innerLoop.getResult(0));
+
+    b.create<InBoundsStoreOp>(loc, loaded, loadBuffer,
+                              outerLoop.getLowerCoords(/*domain*/ 1)[3]);
   }
   return outerLoop;
 }
 
+void copyAndTransposeFromLoadToStoreBuffer(
```

Comment:
I'm not sure I like this approach. Could we, instead, have this `packLoadBufferToStoreBuffer()`, where `loadBuffer` is either `D x K` or `K x D` and `storeBuffer` is `kOuter x D x kpack`?

That way, we'd be calling this code unconditionally and the writes to LDS would appear to be more vectorizable.

Reply:
So the way it's working now is:
- global -> loadBuffer 
- loadBuffer -> storeBuffer
- storeBuffer -> lds

Are you suggesting to have something like:
- global -> loadBuffer
- loadBuffer -> lds
?

Reply:
I'm suggesting this:

```
loadBuffer : KPerThread x DPerThread or DPerThread x KPerThread
storeBuffer : KOuterPerThread x DPerThread x KPackPerThread
lds : KOuter x D x KPack

global memory -> loadBuffer
loadBuffer -[pack]> storeBuffer
storeBuffer -> lds
```

Reply:
I am not sure. I think this is already working like you are proposing. Because we are vectorizing by kpack in the LDS storing routine.

Let's say we have `kperblock=64` , `mperblock=128`, `kpack=16` . `loadBuffer` is already packed into a `16x2` buffer (we load 16 times vectors of length 2) . The final thing we need is to have it transposed in a `2x16` buffer, because we need to store it into the LDS buffer which is `4x128x16`. The fact that `kouter=4` is a workgroup thing, the thread doesn't know it. 

Basically having a private `storeBuffer` sized `4x2x16` wouldn't be useful for the single thread (I think).

Reply:
Ok, let me rephrase, the pack buffer is KPackPerThread, not Kpack - my bad

Reply:
The point is to arrange the data in the order that it'll be sent to LDS by each thread

Reply:
Ok, I applied the changes. I thought of not wrapping the buffers with transforms, but use the transforms directly in the for loop. It's mostly because it seemed too much for a simple transform like this, to have a separate wrapping step.